cats_tutorial.ipynb

```diff
@@ -0,0 +1,491 @@
+    "\n",
+    "Let's recall that in a CB setting, a data point has four components,\n",
+    "\n",
+    "Context\n",
+    "Action\n",
+    "Probability of choosing action\n",
+    "Reward/cost for chosen action\n",
+    "\n",
+    "In our simulator, we will need to generate a context, get an action/decision for the given context and also simulate generating a reward.\n",
+    "\n",
+    "In our simulator, our goal is to maximize reward (TR) or minimize loss (-TR)\n",
+    "\n",
+    "The thermostat tracks two rooms: 'Living Room' and 'Bedroom'\n",
+    "Each room will need temperature adjustment either in the morning or in the afternoon\n",
+    "The context is therefore (room, time_of_day)\n",
+    "\n",
+    "In a continuous range, we can't specify actions since there are infinite actions we can take across the continuous range. We do however provide the minimum value and the maximum value of the range. Here we will range between 0 degrees celsius and 32 degrees celsius which gives us a continuous range of 32 degrees.\n",
+    "\n",
+    "The reward is measured using the absolute difference between the proposed temperature and the one that was actually set by the people living in the house.\n",
+    "\n",
+    "Let's first start with importing the necessary packages:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 57,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import matplotlib.pyplot as plt\n",
+    "from vowpalwabbit import pyvw\n",
+    "import pylibvw\n",
+    "import random\n",
+    "import math\n",
+    "import json"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 58,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# VW tries to minimize loss/cost, therefore we will pass cost as -reward\n",
+    "USER_LIKED_TEMPERATURE = -1.0\n",
+    "USER_DISLIKED_TEMPERATURE = 0.0"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Simulate reward\n",
+    "\n",
+    "In the real world, we will have to learn the room temperature preferences as we observe the interactions between the proposed temperature for each room and the one selected by the people living in the house. Since this is a simulation, we will have to define the preference profile for each room. The reward that we provide to the learner will follow this preference profile. Our hope is to see if the learner can take better and better decisions as we see more samples which in turn means we are maximizing the reward.\n",
+    "\n",
+    "We will also modify the reward function in a few different ways and see if the CB learner picks up the changes. We will compare the TR with and without learning.\n",
+    "\n",
+    "VW optimizes to **minimize cost which is negative of reward**. Therefore, we will always pass negative of reward as cost to VW.\n",
```

Comment:
"VW **minimizes the cost, which is defined as -reward**. Therefore, we will pass the cost associated to each chosen action to VW.\n"